posthog/migrations/0198_property_definition_timestamps.py

```diff
@@ -0,0 +1,77 @@
+# Generated by Django 3.2.5 on 2022-01-19 23:51
+
+import json
+
+import django.utils.timezone
+from django.conf import settings
+from django.db import migrations, models
+from django.db.models import Case, Q, When
+
+from posthog.constants import AnalyticsDBMS
+
+
+def set_created_at(apps, schema_editor):
+    try:
+        from ee.clickhouse.client import sync_execute
+    except ImportError:
+        sync_execute = None  # type: ignore
+
+    EventDefinition = apps.get_model("posthog", "EventDefinition")
+    PropertyDefinition = apps.get_model("posthog", "PropertyDefinition")
+
+    for instance in EventDefinition.objects.all():
+        created_at = None
+        property_keys = []
+        if settings.PRIMARY_DB == AnalyticsDBMS.POSTGRES:
+            Event = apps.get_model("posthog", "Event")
+            # Get first instance of event
+            event_instance = (
+                Event.objects.only("timestamp", "properties")
+                .filter(team=instance.team, event=instance.name)
+                .order_by("timestamp")
+                .first()
+            )
+            if event_instance:
+                created_at = event_instance.timestamp
+                property_keys = list(event_instance.properties.keys())
+        else:
+            if sync_execute:
+                result = sync_execute(
+                    "SELECT timestamp, properties FROM events where team_id=%(team_id)s AND event=%(event)s"
+                    " order by timestamp limit 1",
+                    {"team_id": instance.team.pk, "event": instance.name,},
+                )
+
+            if result:
+                created_at = result[0][0]
+                property_keys = list(json.loads(result[0][1]).keys())
+
+        if created_at and property_keys:
+            # Update all property definitions at once.
+            properties = PropertyDefinition.objects.filter(team=instance.team, name__in=property_keys)
+            properties.update(created_at=Case(When(Q(created_at__lt=created_at) | Q(created_at=None), then=created_at)))
```

Comment:
This feels like something that will cause the migration to fail due to timeouts. Just think of the cloud. Run `select count(*) from posthog_eventdefinition;` and you'll get a 6-digit number. That should be enough to revert this :).

Reply:
Could this be accomplished by an async migration?

Reply:
That's a good point. Async migrations would be a good way to make sure this doesn't timeout.

Also made an optimization that populates a map of property_team keys to created_at values so that PropertyDefinitions can be updated all at once in the end. This should reduce I/O burden and offload it to memory which shouldn't be too bad. As a sanity check I did some load testing on ~100K event definitions... and it froze my M1.

I'll take a stab at rewriting this as an async migration in a less memory + i/o intensive way in the next rev.